cloud/google/cmd/gce-machine-controller/app/controller.go

```diff
@@ -0,0 +1,156 @@
+)
+
+func StartMachineController(server *options.MachineControllerServer, shutdown <-chan struct{}) {
+	config, err := controller.GetConfig(server.CommonConfig.Kubeconfig)
+	if err != nil {
+		glog.Fatalf("Could not create Config for talking to the apiserver: %v", err)
+	}
+
+	client, err := clientset.NewForConfig(config)
+	if err != nil {
+		glog.Fatalf("Could not create client for talking to the apiserver: %v", err)
+	}
+
+	actuator, err := google.NewMachineActuator(server.KubeadmToken, client.ClusterV1alpha1().Machines(corev1.NamespaceDefault), server.MachineSetupConfigsPath)
+	if err != nil {
+		glog.Fatalf("Could not create Google machine actuator: %v", err)
+	}
+
+	si := sharedinformers.NewSharedInformers(config, shutdown)
+	// If this doesn't compile, the code generator probably
+	// overwrote the customized NewMachineController function.
+	c := machine.NewMachineController(config, si, actuator)
+	c.RunAsync(shutdown)
+
+	select {}
+}
+
+func Run(server *options.MachineControllerServer) error {
+	kubeConfig, err := controller.GetConfig(server.CommonConfig.Kubeconfig)
+	if err != nil {
+		glog.Errorf("Could not create Config for talking to the apiserver: %v", err)
+		return err
+	}
+
+	kubeClientControl, err := kubernetes.NewForConfig(
+		rest.AddUserAgent(kubeConfig, "machine-controller-manager"),
+	)
+	if err != nil {
+		glog.Errorf("Invalid API configuration for kubeconfig-control: %v", err)
+		return err
+	}
+
+	recorder, err := createRecorder(kubeClientControl)
+	if err != nil {
+		glog.Errorf("Could not create event recorder : %v", err)
+		return err
+	}
+
+	run := func(stop <-chan struct{}) {
+		StartMachineController(server, stop)
+		panic("unreachable")
+	}
+
+	leaderElectConfig := config.GetLeaderElectionConfig()
+	if !leaderElectConfig.LeaderElect {
+		run(make(<-chan (struct{})))
+		panic("unreachable")
+	}
+
+	// Identity used to distinguish between multiple machine controller instances.
+	id, err := os.Hostname()
+	if err != nil {
+		return err
+	}
+
+	leaderElectionClient := kubernetes.NewForConfigOrDie(rest.AddUserAgent(kubeConfig, "machine-leader-election"))
+
+	id = id + "-" + string(uuid.NewUUID())
+	// Lock required for leader election
+	rl, err := resourcelock.New(
+		leaderElectConfig.ResourceLock,
+		metav1.NamespaceSystem,
+		gceMachineControllerName,
+		leaderElectionClient.CoreV1(),
+		resourcelock.ResourceLockConfig{
+			Identity:      id + "-" + gceMachineControllerName,
+			EventRecorder: recorder,
+		})
+	if err != nil {
+		return err
+	}
+
+	// Try and become the leader and start machine controller loops
+	leaderelection.RunOrDie(leaderelection.LeaderElectionConfig{
+		Lock:          rl,
+		LeaseDuration: leaderElectConfig.LeaseDuration.Duration,
+		RenewDeadline: leaderElectConfig.RenewDeadline.Duration,
+		RetryPeriod:   leaderElectConfig.RetryPeriod.Duration,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: run,
+			OnStoppedLeading: func() {
+				glog.Fatalf("leaderelection lost")
```

Comment:
Is it normal to kill yourself when you lose leadership?

Reply:
Yes, it's pattern how we use leaderelection package throughout k8s. Here is an example.

https://github.com/kubernetes/kubernetes/blob/7ba97b9200a9cab45f59a81ab8b44fca8ee863eb/cmd/cloud-controller-manager/app/controllermanager.go